internal/pkg/api/handleUpload.go

```diff
@@ -0,0 +1,390 @@
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("failed writing error response")
+		}
+		return
+	}
+
+	chunkNum, err := strconv.Atoi(chunkID)
+	if err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("fail writing error response")
+		}
+		return
+	}
+	err = rt.ut.handleUploadChunk(&zlog, w, r, id, chunkNum)
+
+	if err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+
+		// Log this as warn for visibility that limit has been reached.
+		// This allows customers to tune the configuration on detection of threshold.
+		if errors.Is(err, limit.ErrMaxLimit) {
+			resp.Level = zerolog.WarnLevel
+		}
+
+		zlog.WithLevel(resp.Level).
+			Err(err).
+			Int(ECSHTTPResponseCode, resp.StatusCode).
+			Int64(ECSEventDuration, time.Since(start).Nanoseconds()).
+			Msg("fail upload chunk")
+
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("fail writing error response")
+		}
+	}
+}
+
+func (rt Router) handleUploadComplete(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	start := time.Now()
+
+	id := ps.ByName("id")
+
+	reqID := r.Header.Get(logger.HeaderRequestID)
+
+	zlog := log.With().
+		Str(LogAgentID, id).
+		Str(ECSHTTPRequestID, reqID).
+		Logger()
+
+	// simpler authentication check, file integrity checksum
+	// will catch directed tampering, this route just says "im done"
+	if _, err := authAPIKey(r, rt.bulker, rt.ut.cache); err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("failed writing error response")
+		}
+		return
+	}
+
+	err := rt.ut.handleUploadComplete(&zlog, w, r, id)
+
+	if err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+
+		// Log this as warn for visibility that limit has been reached.
+		// This allows customers to tune the configuration on detection of threshold.
+		if errors.Is(err, limit.ErrMaxLimit) {
+			resp.Level = zerolog.WarnLevel
+		}
+
+		zlog.WithLevel(resp.Level).
+			Err(err).
+			Int(ECSHTTPResponseCode, resp.StatusCode).
+			Int64(ECSEventDuration, time.Since(start).Nanoseconds()).
+			Msg("fail upload completion")
+
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("fail writing error response")
+		}
+	}
+}
+
+type UploadT struct {
+	bulker      bulk.Bulk
+	chunkClient *elasticsearch.Client
+	cache       cache.Cache
+	upl         *upload.Uploader
+}
+
+func NewUploadT(cfg *config.Server, bulker bulk.Bulk, chunkClient *elasticsearch.Client, cache cache.Cache) *UploadT {
+	log.Info().
+		Interface("limits", cfg.Limits.ArtifactLimit).
+		Int64("maxFileSize", maxFileSize).
+		Int("maxParallelOps", maxParallelUploadOperations).
+		Int("maxParallelChunks", maxParallelChunks).
+		Msg("Artifact install limits")
```

Comment:
I think this was just copy-pasted from elsewhere and I didn't change the message. Since this is definitely not Artifacts. I guess you've got a capital over there in `handleArtifacts.go`